komorebic/src/main.rs

```diff
@@ -2495,8 +2507,15 @@ Stop-Process -Name:komorebi -ErrorAction SilentlyContinue
         SubCommand::ResizeDelta(arg) => {
             send_message(&SocketMessage::ResizeDelta(arg.pixels))?;
         }
+        // Deprecated
         SubCommand::ToggleWindowContainerBehaviour => {
-            send_message(&SocketMessage::ToggleWindowContainerBehaviour)?;
+            println!("Command deprecated - use new commands 'cycle-window-container-behaviour' or 'window-container-behaviour'");
```

Comment:
I don't think we can do this here; these messages will never be shown to users invoking through whkd/ahk and this will be a de facto breaking change to existing configurations which violates "Breaking changes to user-facing interfaces are unacceptable"

Let's see if we can reformulate this in a way where we can keep `ToggleWindowContainerBehaviour` as it is, and maybe introduce a new toggle which turns the start-as-floating behaviour on and off.

Reply:
I'm not too fussy on the under-the-hood implementation details as long as this command (which I know has quite a few die-hard users) remains in-tact

Reply:
I don't think it makes much sense to have a `ToggleWindowContainerBehaviour` when there are more than two states. Toggle makes you think there are only two states which is not true, but thats just my opinion. I've added two new commands `WindowContainerBehaviour` which lets you set a specific behaviour and `CycleWindowContainerBehaviour` which lets you cycle behaviours on a certain direction.

I could keep the `ToggleWindowContainerBehaviour` but make it act as `CycleWindowContainerBehavior next` which would be similar to what it does right now with the added `Float` behaviour.

But if you feel it is really important that the old command keeps doing the same thing I can do that as well. However what should that command do if the current set behaviour is `Float`? Should it change it to `Create`?

This might become confusing, we might need to add the comment below so we don't forget in future why does this look like this:
```rust
            SocketMessage::ToggleWindowContainerBehaviour => {
                // This command was kept as it was before adding the `Float` variant
                // for backwards compatibility. It toggles between `Create` and
                // `Append`, unless it is on `Float` where it sets it to `Create`.
                match self.window_container_behaviour {
                    WindowContainerBehaviour::Create => {
                        self.window_container_behaviour = WindowContainerBehaviour::Append;
                    }
                    WindowContainerBehaviour::Append => {
                        self.window_container_behaviour = WindowContainerBehaviour::Create;
                    }
                    WindowContainerBehaviour::Float => {
                        self.window_container_behaviour = WindowContainerBehaviour::Create;
                    }
                }
            }
```

Tell me which one do you prefer:
1) Keep the `ToggleWindowContainerBehaviour` but make it act as `CycleWindowContainerBehaviour next` (basically I would change the `Append` arm above to return `Float` instead of `Create`)
2) Keep the `ToggleWindowContainerBehaviour` as shown above and users can use `WindowContainerBehaviour` to set to `Float` if they need/want
3) Keep the `ToggleWindowContainerBehaviour` as shown above and add `ToggleFloatWindowContainerBehaviour` to toggle the behaviour between "Float" and "previous state" (this will require keeping track of the previous state)
4) Keep the `ToggleWindowContainerBehaviour` as shown above and add `ToggleFloatWindowContainerBehaviour` to toggle the behaviour between "Float" and "Create" (this simpler to implement, but users might not like it...)

Reply:
It's also possible to do nothing if `::Float` is set and we stick with expanding the `WindowContainerBehaviour` enum:

```rust
SocketMessage::ToggleWindowContainerBehaviour => {
  match self.window_container_behaviour {
      WindowContainerBehaviour::Create => {
          self.window_container_behaviour = WindowContainerBehaviour::Append;
      }
      WindowContainerBehaviour::Append => {
          self.window_container_behaviour = WindowContainerBehaviour::Create;
      }
      WindowContainerBehaviour::Float => // do nothing
  }
 }
```

> to toggle the behaviour between "Float" and "previous state" (this will require keeping track of the previous state)

We can investigate a separate boolean toggle on the `WindowManager` struct, or wrapping `WindowContainerBehaviour` in a higher level struct which can maintain previous state:

```rust
pub struct WindowManagementBehaviour {
    container_behaviour: WindowContainerBehaviour,
    floating_override: bool,
}
```

The way I see this option is more as an override; regardless of what other behaviours are set for new windows, when this is activated it always takes precedence.

... and now that I'm thinking of this as an override, it should probably live at the `Workspace` level rather than the `WindowManager` level, because every layout-related option that I've added at the `WindowManager` or `Monitor` level has always had a request to push the configurability down to the `Workspace` level within a day of release 😅

Reply:
> because every layout-related option that I've added at the `WindowManager` or `Monitor` level has always had a request to push the configurability down to the `Workspace` level within a day of release 😅

Yeah I've already asked that myself [here](https://github.com/LGUG2Z/komorebi/pull/1006#issuecomment-2382636596)! 😅 

Ok. So if you want to I can refactor this to use this override approach with `WindowManagementBehaviour` struct. Then the commands would be easy, it would be the old one and a new one to override float, something like `ToggleWindowContainerBehaviourFloatOverride` or is that too many words, maybe `ToggleFloatOverride`.

Now I have some more questions:
1) Should it be only on workspace, or should there be a global behaviour that workspaces can override, kind of like the workspace padding?
2) Should the commands only work on the currently focused workspace or should there be multiple commands for currently focused workspace, for workspace number, for workspace name and for monitor/workspace numbers? If you want commands for all these cases I'll need help with the names:
 - `ToggleWindowContainerBehaviour`: toggles global behaviour
 - `ToggleWindowContainerBehaviourWorkspace`: toggles currently focused workspace behaviour (if not set yet, uses opposite of global one)
 - `ToggleWindowContainerBehaviourWorkspaceNumber(usize)`: toggles workspace behaviour for the specified number (if not set yet, uses opposite of global one)
 - `ToggleWindowContainerBehaviourNamedWorkspace(String)`: toggles workspace behaviour for the specified named workspace (if not set yet, uses opposite of global one)
 - `ToggleWindowContainerBehaviourMonitorWorkspaceNumber(usize, usize)`: toggles workspace behaviour for the specified monitor and workspace numbers combination (if not set yet, uses opposite of global one)
 And then the same thing for the `ToggleFloatOverride` (or other name if you prefer...)

Reply:
> Should it be only on workspace, or should there be a global behaviour that workspaces can override, kind of like the workspace padding?

I think this is a good established pattern that we can go with

> Should the commands only work on the currently focused workspace or should there be multiple commands for currently focused workspace, for workspace number, for workspace name and for monitor/workspace numbers? If you want commands for all these cases I'll need help with the names:

Let's do currently focused workspace for now, I think the global + workspace granularity should cover 99% of use cases since usually this is something that people want to toggle on the current workspace

For the socket message / command naming we should be good with:

`ToggleFloatOverride`
`ToggleWorkspaceFloatOverride`
`ToggleWindowContainerBehaviour`
`ToggleWorkspaceWindowContainerBehaviour`